### PR TITLE
Improvements to rewriting systems and 2-cohomology

### PR DESCRIPTION
**Pull request by hulpke**
Improvements to stored RWS data from hand-coded BN-systems
2-cohomology reduces equations as it goes
Added manual warning about pc group generation through RWS

### CODE REVIEW

File: lib/gpfpiso.gi [hunk 1]
```diff
@@ -1557,11 +1566,615 @@ end);
...
+        Add(a,-2*i); 
+      fi;
+    od;
+    a:=AssocWordByLetterRep(FamilyObj(One(f)),a);
+    return a;
+  end;
+
+  trawou:=w->trawo(UnderlyingElement(w));
+
+  lev:=Concatenation(lev,
```
Comment (fingolfin) [inline lib/gpfpiso.gi:1946]:
```suggestion
  Append(lev,
```

File: lib/gpfpiso.gi [hunk 2]
```diff
@@ -1557,11 +1566,615 @@ end);
...
+        k:=List(jj,x->trawou(borelword(x)));
+        b:=a[1]*Product([1..Length(jj)],x->k[x]*a[x+1]);
+        j:=j[1]*j[2]*j[3];
+        b:=[b,trawou(j)];
+        addrule(b);
+      fi;
+    od;
+  od;
+
+  # group relators to make sure its the proper group
```
Comment (fingolfin) [inline lib/gpfpiso.gi:2066]:
```suggestion
  # group relators to make sure it's the proper group
```

File: lib/grppcext.gd [hunk 3]
```diff
@@ -130,6 +130,9 @@ DeclareGlobalFunction( "EXReducePermutationActionPairs" );
 ##  The optional argument <A>D</A> should be a subgroup of
 ##  Aut(<A>G</A>) <M>\times</M> Aut(<A>M</A>).
 ##  If it is given, then only the compatible pairs in <A>D</A> are computed.
+##  If a group <A>A</A> of automorphisms of <A>G</A> is given as optional
+##  first argument, it is used in place of the full automorphism group of
+##  <A>G</A>, avoiding the need to copmpute this automorphism group.
```
Comment (fingolfin) [inline lib/grppcext.gd:135]:
```suggestion
##  <A>G</A>, avoiding the need to compute this automorphism group.
```

File: lib/twocohom.gi [hunk 4]
```diff
@@ -1001,36 +1002,45 @@ local field,fp,fpg,gens,hom,mats,fm,mon,kb,tzrules,dim,rules,eqs,i,j,k,l,o,l1,
...
-#          else
-#            Print(">>Overlap ",l1," ",l2," ",o," does not reduce further\n");
-#          fi;
-
         fi;
       od;
     od;
+    Add(olen,Length(BasisVectors(eqs)));
+    if Length(olen)>3 and 
+      # if twice stayed teh same after increase
```
Comment (fingolfin) [inline lib/twocohom.gi:1026]:
```suggestion
      # if twice stayed the same after increase
```

**Review by fingolfin — APPROVED** [review]
Looks good to me. Some minor and optional remarks added. One question (again, not a blocker, just to make sure I understood things right)

File: grp/simplerew.grp [hunk 5]
```diff
@@ -16,185 +16,6 @@
 
 return
 [
-[[20160,"L",[3,4],"L3(4)"],
```
[the remark this answers is not in the document]
Comment (hulpke) [reply grp/simplerew.grp:19]: Yes. For Type L the computation now is quick and good that it is not worth to precompute. (Ultimately this should hold for all classical groups.)

File: lib/gpfpiso.gi [hunk 6]
```diff
@@ -918,8 +917,9 @@ InstallMethod( IsomorphismFpGroupBySubnormalSeries,
 function( G, series, str )
     local l, H, gensH, iso, F, gensF, imgsF, relatorsF, free, n, k, N, M,
           hom, preiH, c, new, T, gensT, E, gensE, imgsE, relatorsE, rel,
-          w, t, i, j;
+          w, t, i, j,known;
 
+    known:=ValueOption("knownfactor");
```
[the remark this answers is not in the document]
Comment (hulpke) [reply lib/gpfpiso.gi:922]: It is called in `SplitBNRewritingPresentation` to ensure the presentation exposes a Weyl group factor. I can add test examples, but they take a bit more time, in particular if assertions are on.

**Pull request merged**